### PR TITLE
Fix --debug flag

### PR DESCRIPTION
**Pull request by BrennanConroy**
#1418

### CODE REVIEW

**Comment by BrennanConroy** [comment]
@davidfowl @victorhurdugaci

File: src/dnx/dnx.cpp [hunk 1]
```diff
@@ -391,6 +391,14 @@ int wmain(int argc, wchar_t* argv[])
                 }
 
                 ::wprintf_s(L"Debugger attached.\r\n");
+
+                //remove --debug flag from arguments
+                if (argc > i)
```
Comment (BrennanConroy) [inline src/dnx/dnx.cpp]: Should be argc > i + 1
Also, need to move the argv[i] = L"\0"; to outside of the if check

File: src/dnx/dnx.cpp [hunk 2]
```diff
@@ -391,6 +391,14 @@ int wmain(int argc, wchar_t* argv[])
...
+                {
+                    memcpy(argv + i, argv + i + 1, (argc - i) * sizeof(wchar_t));
+                    argv[i] = L"\0";
```
Comment (BrennanConroy) [inline src/dnx/dnx.cpp]: argv[argc - 1] = L"\0";

**Comment by ellismg** [comment]
FYI: I will need to rebase my dnx changes after this goes in.

**Comment by victorhurdugaci** [comment]
Looks good :shipit: (after you rebase)

**Comment by davidfowl** [comment]
Don't like this fix. Handle it like appbase. Consume the argument in managed code.

**Comment by anurse** [comment]
That makes the experience inconsistent when using batch/shell scripts like "kpm". If I run: `kpm --debug restore`, and the AppHost is the one responsible for removing the arg, it doesn't work:

1. DNX.exe waits for the debugger
2. AppHost gets "dnx --appbase ... Microsoft.Framework.ApplicationHost Microsoft.Framework.PackageManager --debug restore"
3. AppHost doesn't see the "--debug" arg so it flows through to KPM, which has to know about it.

IF we're going to support --debug anywhere (which the current code does), then the native code has to strip it. The alternative would be to only support `--debug` prior to the actual command (which means `kpm --debug restore` would NOT trigger the debug behavior unless kpm ALSO has a debug switch of it's own)

**Comment by davidfowl** [comment]
It doesn't because app host isn't removing the flag. Look at how appbase is handled and do the same

**Comment by anurse** [comment]
We can handle it the same way, but we have to change the native code that reads it to only accept it as an argument TO apphost. If it ends up somewhere AFTER the app command name, apphost ignores it and passes it along (yes, EVEN when you register the option in AppHost's ParseArgs). 

Right now, app host options can only appear BEFORE the name of the DLL to run because CommandLineOptions stops parsing Options after the first Argument appears.

**Comment by davidfowl** [comment]
Ok, to be clear. Undo how the debug flag works today and treat this flag light --appbase. Look at the existing code and do whatever appbase does. Today --appbase is a flag that is parsed in native code and consumed by the first set of managed code before the app runs.

**Comment by BrennanConroy** [comment]
Updated

**Comment by anurse** [comment]
To truely be like `--appbase`, we also need to change the way the `--debug` flag is detected and acted on in native code. Right now, if `--debug` appears ANYWHERE in the command line, we wait for the debugger, but we'll only strip it out if it appears BEFORE the project.json path (because AppHost strips it out).

We should change the native code to **only** support the `--debug` switch **before** the project.json path. Individual K apps are responsible for setting up their own debug hook if they want it (i.e. `kpm --debug` would be handled in `kpm`)

**Comment by BrennanConroy** [comment]
@anurse @davidfowl

**Comment by anurse** [comment]
Looks good now :). :shipit:

File: src/dnx/dnx.cpp [hunk 3]
```diff
@@ -375,8 +375,17 @@ int CallApplicationProcessMain(int argc, wchar_t* argv[])
 int wmain(int argc, wchar_t* argv[])
 {
     // Check for the debug flag before doing anything else
-    for (int i = 0; i < argc; i++)
+    for (int i = 1; i < argc; i++)
     {
+        //anything without - or -- is appbase or non-dnx command
+        if (::_wcsnicmp(argv[i], L"-", 1) != 0)
```
Comment (moozzyk) [inline src/dnx/dnx.cpp]: why not just `argv[I][0] == L'-'` ?

File: src/dnx/dnx.cpp [hunk 4]
```diff
@@ -375,8 +375,17 @@ int CallApplicationProcessMain(int argc, wchar_t* argv[])
...
+        {
+            break;
+        }
+        if (StringsEqual(argv[i], L"--appbase"))
```
Comment (moozzyk) [inline src/dnx/dnx.cpp]: These two could be nested - if it does not start wit - it won't start with --

Comment (moozzyk) [inline src/dnx/dnx.cpp]: Forget it

File: src/dnx.mono.managed/EntryPoint.cs [hunk 5]
```diff
@@ -16,9 +16,27 @@ public class EntryPoint
...
-        bool hasDebugWaitFlag = arguments.Any(arg => string.Equals(arg, "--debug", StringComparison.OrdinalIgnoreCase));
+        bool hasDebugWaitFlag = false;
+        foreach (var arg in arguments)
+        {
+            //anything without - or -- is appbase or non-dnx command
+            if (arg.ElementAt(0) != "-")
+            {
+                break;
+            }
+            if(arg == "--appbase")
```
Comment (victorhurdugaci) [inline src/dnx.mono.managed/EntryPoint.cs]: not needed

File: src/dnx.mono.managed/EntryPoint.cs [hunk 6]
```diff
@@ -16,9 +16,27 @@ public class EntryPoint
     public static int Main(string[] arguments)
     {
         // Check for the debug flag before doing anything else
-        bool hasDebugWaitFlag = arguments.Any(arg => string.Equals(arg, "--debug", StringComparison.OrdinalIgnoreCase));
+        bool hasDebugWaitFlag = false;
+        foreach (var arg in arguments)
+        {
+            //anything without - or -- is appbase or non-dnx command
+            if (arg.ElementAt(0) != "-")
```
Comment (victorhurdugaci) [inline src/dnx.mono.managed/EntryPoint.cs]: ugh... not `ElementAt`. just use the indexer `[0]`

File: src/dnx.mono.managed/EntryPoint.cs (hunk 5 above)
Comment (victorhurdugaci) [inline src/dnx.mono.managed/EntryPoint.cs]: Wait a second... why do you stop after `appbase`?

**Comment by ellismg** [comment]
Please rebase on top of 24dd1f8d and fix up the merge conflicts.  You'll need to adopt the _T() macros in dnx.cpp and I think I moved some code you touched into the PAL.  Please let me know if you need help fixing things up.

**Comment by BrennanConroy** [comment]
@ellismg rebased, could you take a quick look to see if I missed anything?

File: src/dnx.host/RuntimeBootstrapper.cs [hunk 7]
```diff
@@ -65,6 +65,7 @@ public static Task<int> ExecuteAsync(string[] args)
             // We just need to capture them and clean them up.
             var optionAppbase = app.Option("--appbase <PATH>", "Application base directory path",
                 CommandOptionType.SingleValue);
+            var optionDebug = app.Option("--debug", "Wait for debugger to attach", CommandOptionType.NoValue);
```
Comment (BrennanConroy) [inline src/dnx.host/RuntimeBootstrapper.cs]: Looks like someone else added this, will remove

**Comment by ellismg** [comment]
@BrennanConroy LGTM

File: src/dnx.mono.managed/EntryPoint.cs [hunk 8]
```diff
@@ -16,9 +16,29 @@ public class EntryPoint
...
-        bool hasDebugWaitFlag = arguments.Any(arg => string.Equals(arg, "--debug", StringComparison.OrdinalIgnoreCase));
+        bool hasDebugWaitFlag = false;
+        for (var i = 0; i < arguments.Length; ++i)
+        {
+            //anything without - or -- is appbase or non-dnx command
+            if (arguments[i][0] != '-')
+            {
+                break;
+            }
+            if(arguments[i] == "--appbase")
```
Comment (davidfowl) [inline src/dnx.mono.managed/EntryPoint.cs]: spacing!

**Comment by davidfowl** [comment]
Did you add this to the *nix PAL as well?

**Comment by BrennanConroy** [comment]
Change is only in dnx.cpp. No extra work needed.

**Comment by davidfowl** [comment]
:shipit:

**Pull request closed without merging**